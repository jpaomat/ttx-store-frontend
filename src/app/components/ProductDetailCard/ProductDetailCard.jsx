import { Button } from '../../shared/ui/components/Button/Button';
import './ProductDetailCard.scss';

export const ProductDetailCard = () => {
	const data = {
		brand: 'Acer',
		model: 'Iconia Talk S',
		price: '784',
		imgUrl:
			'https://itx-frontend-test.onrender.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg',
		cpu: 'Quad-core 1.3 GHz Cortex-A53',
		ram: '2 GB RAM',
		os: 'Android 6.0 (Marshmallow)',
		displaySize: '720 x 1280 pixels (~210 ppi pixel density)',
		battery: 'Non-removable Li-Ion 3400 mAh battery (12.92 Wh)',
		dimentions: '191.7 x 101 x 9.4 mm (7.55 x 3.98 x 0.37 in)',
		primaryCamera: ['13 MP', 'autofocus'],
		secondaryCmera: ['2 MP', '720p'],
        options: {
            colors: [
                {
                    code: 1000,
                    name: 'Black'
                }
            ],
            storages: [
                {
                    code: 2000,
                    name: '16 GB'
                },
                {
                    code: 2001,
                    name: '32 GB'
                }
            ]
        },
		colors: [
			'Black',
			'White'
		],
	};

	const dataKeys = [
		'brand',
		'model',
		'price',
		'cpu',
		'ram',
		'os',
		'displaySize',
		'battery',
		'dimentions',
		'primaryCamera',
		'secondaryCmera',
	];

	return (
		<div className='detail-card'>
			<div className='detail-card__item detail-card__item--image'>
				<img src={data.imgUrl} alt='' />
			</div>
			<div className='detail-card__item item-information'>
				<div className='item-information__header'>
					<p className='item-information__header--title'>
						{`${data.brand}/${data.model}/${data.ram}`}
					</p>
					<p className='item-information__header--price'>
					</p>
					{data.price ?
						<p className='item-information__header--price'>{data.price} â‚¬</p> :
						<p className='item__no-stock '>No disponible</p>
					}
				</div>

				<div className='item-information__body'>
					<p className='item-information__body--title'>Caracteristicas destacadas</p>
					<ul className='item-information__body--options'>
						{dataKeys.map(key => (
							<li key={key}>
                                <span><b>{key}: </b> {data[key]}</span>
							</li>
						))}
					</ul>
				</div>

				<div className='item-information__footer'>
					<div className='item-information__footer--actions'>
                        {['Almacenamiento', 'Color', 'Cantidad'].map((action, k) => (
                            <div className='select-actions' key={k}>
                                <label className='select-actions__label'>{action}</label>
                                <select className='select-actions__select' id="actions" name="actions">
                                    {data.options.storages.map((storage, key) =>
                                        <option key={key} value={storage.name}>{storage.name}</option>
                                    )}
                                </select>
                            </div>
                        ))}
                    </div>

					<Button size='sm' disabledBtn={!data.price}>
						<span>{ data.price ? 'Agregar al carro' : 'No disponible'}</span>
					</Button>
				</div>
			</div>
		</div>
	);
};
