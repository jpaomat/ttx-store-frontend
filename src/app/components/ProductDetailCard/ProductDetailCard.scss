@import '../../../assets/scss/';
.detail-card {
	&__item {
		margin: 24px;
		&--image {
			display: flex;
			img {
				width: 100%;
				min-width: 100px;
				max-width: 280px;
				margin: auto;
			}
		}
	}
}

.item-information {
	&__header {
		font-size: 14px;

		&--title {
			font-size: 20px;
		}
		&--price {
			font-size: 24px;
			font-weight: 600;
			color: $purple-500;
			line-height: 64px;
		}
	}

	&__body {
		line-height: 24px;
		&--title {
			font-weight: bold;
		}
		&--options {
			padding: 10px;
			font-size: 14px;
		}
	}

	&__footer {
		&--actions {
			font-size: 12px;
			margin: 10px 0 24px 0;

			.select-actions {
				margin-top: 12px;
				&__label {
					margin: 4px;
				}
				&__select {
					border: 1px solid #7028fc;
					color: #7028fc;
				}
			}
		}
	}
}

@media screen and (min-width: $break-desktop) {
	.detail-card {
		@include organizedMultipleColumns(40% 60%);
		margin: 48px;

		.item-information {
			border-left: 1px solid $gray-220;
			// border-radius: 8px;
			padding: 8px 48px;

			&__header {
				&--title {
					font-size: 28px;
				}
				&--price {
					font-size: 2rem;
				}
			}

			&__body {
				line-height: 24px;
				&--title {
					font-weight: bold;
				}
				&--options {
					margin: 0 24px 12px;
				}
			}

			&__footer {
				&--actions {
					// @include organizedMultipleColumns(repeat(3, 1fr));
    				display: flex;
					justify-content: space-evenly;
					margin: 10px 0 32px 0;
					.select-actions {
						&__label {
							margin: 4px;
						}
					}
				}
			}
		}
	}
}
