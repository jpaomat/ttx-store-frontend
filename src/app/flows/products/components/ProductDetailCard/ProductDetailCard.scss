@import '../../../../../assets/scss/';

.detail-card {
	background-color: $white;
	&__item {
		margin: 24px;
		&--image {
			display: flex;
			border: 1px solid $purple-100;
			img {
				width: 100%;
				min-width: 100px;
				max-width: 280px;
				margin: auto;
			}
		}
	}
}

.item-information {
	&__header {
		font-size: 14px;

		&--title {
			font-size: 20px;
		}
		.item-price {
			color: $orange-500;
			font-size: 32px;
		}
		&--price {
			font-weight: 600;
			line-height: 64px;
		}
	}

	&__body {
		line-height: 24px;
		&--title {
			font-weight: bold;
		}
		&--options {
			padding: 10px;
			font-size: 14px;
		}
	}

	&__footer {
		&--actions {
			font-size: 12px;
			margin: 10px 0 24px 0;
		}
	}
}

@media screen and (min-width: $break-tablet) {
	.detail-card {
		padding: 12px;
		.item-information {
			&__footer {
				padding: 0 10%;
			}
		}
	}
}

@media screen and (min-width: $break-desktop) {
	.detail-card {
		@include organizedMultipleColumns(45% 55%);
		margin: 18px 48px;
	}
	.item-information {
		border-radius: 10px;

		&__header {
			&--title {
				font-size: 28px;
			}
			&--price {
				font-size: 48px;
			}
			&--currrency {
				font-size: 32px;
			}
		}

		&__body {
			line-height: 18px;
			&--title {
				font-weight: bold;
			}
			&--options {
				margin: 0 24px 12px;
			}
		}

		&__footer {
			padding: 0 5%;
			&--actions {
				@include organizedMultipleColumns(50% 50%);
			}
		}
	}
}

