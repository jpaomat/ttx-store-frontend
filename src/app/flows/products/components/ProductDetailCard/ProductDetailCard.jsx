import { useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import { useState } from 'react';
import { BsCart4 } from 'react-icons/bs';
import PropTypes from 'prop-types';
import { ButtonApp, MultiselectContainer } from '../../../../shared/ui/components';
import {
	dataToShowDescription,
	initializedProductSelected,
	selectorsProduct,
} from '../../const/detailsData.const';
import { incrementCartCounter } from '../../../../store/slices/cartCounter/cartCounterSlice';
import './ProductDetailCard.scss';
import { productsApi } from '../../../../../api/products';

export const ProductDetailCard = ({
	productSelected = initializedProductSelected,
}) => {
	const { productId } = useParams();
	const [selectedProductOptions, setSelectedProductOptions] = useState(null);
	const [selectorDefaultValues, setSelectorDefaultValues] = useState(null);
	const dispatch = useDispatch();

	const getOptions = (codeName, options) => {
		return options.map(({ code, name }) => ({ codeName, code, name }));
	};

	const onSelect = optionSelected => {
		const { codeName, code } = optionSelected[0];
		setSelectedProductOptions({
			...selectorDefaultValues,
			...selectedProductOptions,
			...{ [`${codeName}`]: code },
		});
	};

	const addItemCart = () => {
		let dataToSend = selectedProductOptions || selectorDefaultValues;
		dataToSend = { ...dataToSend, id: productId };
		productsApi
			.post('cart', {...dataToSend})
			.then(({ data }) => dispatch(incrementCartCounter({ ...data, product: dataToSend })))
			.catch(error => console.error(error));
	};

	// if (!productSelected) {
	// 	return <Navigate to='/products' />;
	// }

	const { brand, imgUrl, model, ram, price, options } = productSelected;

	return (
		productSelected.brand && (
			<div className='detail-card'>
				<div className='detail-card__item detail-card__item--image'>
					<img src={imgUrl} alt='' />
				</div>
				<div className='detail-card__item item-information'>
					<div className='item-information__header'>
						<p className='item-information__header--title'>
							{`${brand}/${model}/${ram}`}
						</p>
						{price ? (
							<div className='item-price'>
								<span className='item-information__header--price'>{price}</span>
								<span className='item-information__header--currrency'>â‚¬</span>
							</div>
						) : (
							<p className='item__no-stock '>No disponible</p>
						)}
					</div>

					<div className='item-information__body'>
						<p className='item-information__body--title'>
							Caracteristicas destacadas
						</p>
						<ul className='item-information__body--options'>
							{dataToShowDescription.map(({ label, property }) => (
								<li key={property}>
									<span>
										<b>{label}: </b> {productSelected[property]}
									</span>
								</li>
							))}
						</ul>
					</div>

					<div className='item-information__footer'>
						<div className='item-information__footer--actions'>
							{selectorsProduct.map(({ selector, label, codeName }) => (
								<MultiselectContainer
									key={selector}
									selector={selector}
									label={label}
									displayValue='name'
									selectedValues={getOptions(codeName, options[selector])[0]}
									setSelectorDefaultValues={setSelectorDefaultValues}
									options={getOptions(codeName, options[selector])}
									onSelect={onSelect}
								/>
							))}
						</div>
						<ButtonApp size='sm' disabledBtn={!price} onClick={addItemCart}>
							{price ? (
								<>
									<span>Agregar al carrito</span>
									<BsCart4 className='cart-icon' size={14} />
								</>
							) : (
								<span>No disponible</span>
							)}
						</ButtonApp>
					</div>
				</div>
			</div>
		)
	);
};

ProductDetailCard.defaultProps = {
	productSelected: {},
};

ProductDetailCard.propTypes = {
	productSelected: PropTypes.shape({}),
};
