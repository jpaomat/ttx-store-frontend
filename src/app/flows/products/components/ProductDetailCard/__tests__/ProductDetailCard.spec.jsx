/* eslint-disable no-unused-vars */
import Renderer from 'react-test-renderer/shallow';
import axios from 'axios';
import { Provider } from 'react-redux';
import { ProductDetailCard } from '../ProductDetailCard';

const renderer = new Renderer();

jest.mock('axios', () => ({
  post: jest.fn(),
  get: jest.fn(),
  create: jest.fn()
}));

describe('Product detail wrap component', () => { 
    it('Renderer componen', ()=> {
        // const component = renderer.render(
        //     <Provider>
        //         <ProductDetailCard />
        //     </Provider>
        // );
        // expect(component).toMatchSnapshot();
    });
 });