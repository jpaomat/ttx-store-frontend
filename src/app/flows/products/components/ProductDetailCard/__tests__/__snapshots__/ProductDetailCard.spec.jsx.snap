// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product detail wrap component Renderer componen 1`] = `undefined`;
